Multilevel coarsening with more that one coarse dof per coarse entity

chakshinglee opened the pull request:
This PR pretty much completes the necessary components for generating "high order" coarse spaces recursively. 

The key change is to aggregate dofs instead of vertices/edges during coarsening step, which is not too difficult with the help of `GraphSpace`. 

One additional thing that we need is `AggregateDof`, which tells what are the finer level vdofs or evofs associated with some aggregate or face (coarse entity). They are needed when computing targets and solving extension/bubble problems. And it is a generalization of `GraphTopology`. I tried to keep the addition as minimal as possible, so just make `AggregateDof` as a `struct` storing all those relation tables to make the interface clean.

The rest of the core changes is to compute extended aggregate to dofs (instead of entity), and and use the aggregate to dof (instead of entity) for local matrix extraction.

Some supplementary additions are the sparse eigensolver for non-diagonal M (dense solver can be pretty slow for high order) and projection from the coarse vertex space to piecewise constant (on aggregate), and the inverse mapping. The latter two are used in `PDESampler` when projecting the coarse solution to a piecewise constant function and so do some (nonlinear) function evaluation and then project back to the coarse vertex space.

One additional tests for ml higher order space is ml-sampler4 (using `max_evects = 4`). And the test on `coarse_assembling` is set to higher order `max_evects  = 3" as well, verifying the coarse matrices are same as doing RAP on fine matrices.


Edited: my bad that in the above when I said `AggregateDof` I actually meant `DofAggregate`.

chakshinglee committed:
modulize some common init process and add some access functions in Graph

chakshinglee committed:
remove some unnecessary members (starts arrays) in GraphTopology

chakshinglee committed:
some simplification in BuildExtendedAggregate and GraphTopology

chakshinglee committed:
fix some wrong starts array in lineargraph

chakshinglee committed:
replace usage of LeftDiagMult by the equivalent ParMult function.

chakshinglee committed:
Merge branch 'feature/graphspace' of https://github.com/LLNL/smoothG into feature/ml-high-order

chakshinglee committed:
use coarse entity to fine dof relation instead of fine entity in computation of local targets

chakshinglee committed:
use entity to dofs in CollectConstant

chakshinglee committed:
use dofs notion in CombineM, ComputeEdgeTargets and BuildPvertices

chakshinglee committed:
use dofs notion in BuildPedges

chakshinglee committed:
use dofs notion in CoarseMBuilder

chakshinglee committed:
simplifying InitializePEdgesNNZ

chakshinglee committed:
use dofs notion in BuildAggregateFaceM

chakshinglee committed:
make style

chakshinglee committed:
simplifying CoarseMBuilder using GraphSpace

chakshinglee committed:
remove Agg_edge_ from GraphTopology because it is no longer needed (in multilevel 'agg_edof' is needed instead)

chakshinglee committed:
remove Agg_edge_ in lineargraph

chakshinglee committed:
further simplifying CoarseMBuilder

chakshinglee committed:
fix wrong offsets

chakshinglee committed:
fix coarse M value in 'higher order' coarsening

chakshinglee committed:
use first vertex target in each aggregate as local constant representation (instead of use subvector of global constant_rep). turns out this improves the precision of resulting coarse M (see the test coarse_assembling)

chakshinglee committed:
add a struct (DofAggregate) containing dof aggregation information. This avoid repeated computation of those information during coarsening

chakshinglee committed:
update documentation, simplify LocalEdgeSolver

chakshinglee committed:
add a local eigensolver for the case of general M that use ARPACK (by providing the action of A inverse through block solve)

chakshinglee committed:
correct the trace computation in the case of non-diagonal M (M inverse time D^T evects)

chakshinglee committed:
put the implementation of eigensolver for general M into LocalEigenSolver regardless dense or sparse matrix format (just like primal form)

chakshinglee committed:
refactor interface of MixedBlockEigensystem, avoid one copy in ComputeEigenvectors (not substantial)

chakshinglee committed:
make style

chakshinglee committed:
resolve conflicts from merging feature/graphspace

chakshinglee committed:
Merge branch 'master' of https://github.com/LLNL/smoothG into feature/ml-high-order

chakshinglee committed:
test higher order coarsening (m=3) in coarse_assembling

chakshinglee committed:
unify two loops as lambda

chakshinglee committed:
add preprocessor flag to suppress compiler warning

chakshinglee committed:
fix unbalance tolerance that rendered an empty graph on a processor

chakshinglee committed:
add a piecewise-constant projector in MixedMatrix

chakshinglee committed:
resolve conflicts from merging feature/graphspace

chakshinglee committed:
use piece-wise constant projector in Sampler

chakshinglee committed:
use piece-wise constant interpolator in Sampler

chakshinglee committed:
adding a test on multilevel sampler that uses higher order coarse space (m = 4)

barker29 commented on examples/stest.py:
```diff
@@ -649,10 +649,21 @@ def make_tests():
          {"p-error-level-1": 0.20833920382939719,
           "p-error-level-2": 0.40118441747952621}]
 
+    tests["ml-sampler4"] = \
```
This test doesn't run because it's not added to `examples/CMakeLists.txt` (around line 125). (we should maybe automate this better so you only have to add the test in one place).

chakshinglee committed:
add test in CMakeLists.txt

barker29 commented on src/MatrixUtilities.hpp:
```diff
@@ -79,6 +79,17 @@ mfem::SparseMatrix AAt(const mfem::SparseMatrix& A);
+/**
+    @brief Compute the product A * B between SparseMatrix and HypreParMatrix
+*/
+std::unique_ptr<mfem::HypreParMatrix> ParMult(const mfem::HypreParMatrix& A,
```
I'm having a hard time understanding this function. Am I right in understanding that the `SparseMatrix` object has different data on each processor, this routine just reinterprets that data as a parallel object? Why don't we just store it as a `HypreParMatrix` all the time?

chakshinglee replied:
You are right. This function will interpret a `SparseMatrix` as a "block-diagonal" `HypreParMatrix` and then call `mfem::ParMult`. This kind of multiplication is used in several places (e.g. `GraphTopology`, `LocalMixedGraphSpectralTargets`), so I just make the process (actually just a few lines) as a function. I will add more details in the documentation.

For the question whether we should just store the `SparseMatrix` as a `HypreParMatrix`, I think most of the time when this function is called, the `SparseMatrix` object is either temporary or it is more convenient to be in that format (e.g. `MixedMatrix::D_` for local extraction). So changing the storage format may result in more changes.

barker29 replied:
That makes sense. Your solution is fine with me.

barker29 commented on src/MixedMatrix.hpp:
```diff
@@ -69,7 +69,9 @@ class MixedMatrix
                 std::unique_ptr<mfem::SparseMatrix> W,
-                mfem::Vector constant_rep);
+                mfem::Vector constant_rep,
+                mfem::SparseMatrix Ppw1,
```
Can we document `Ppw1` and `Qpw1` here? (the names are not helpful unless you already know...)

barker29 commented on src/Upscale.hpp:
```diff
@@ -140,15 +140,16 @@ class Upscale : public mfem::Operator
         return GetMatrix(level).GetConstantRep();
     }
 
+    mfem::Vector PWConstProject(int level, const mfem::Vector& x) const;
```
A line of documentation please (you can just copy the comments from `MixedMatrix.hpp` lines 373-374).

barker29 reviewed (commented):
This looks very good!

In addition to my comments on a few lines of code, I have two other comments:
- I'd like to see a three-level spectral test with `generalgraph` or `finitevolume`.
- For me, the `pargraph-usegenerator` test is failing (looks like a minor difference in operator complexity)

barker29 committed:
bugfix: in SetBubbleBubbleBlock, take agg number as argument

This fixes errors in case where RegisterRow is not called and
agg_index_ is not set in the coarse_m_builder_ object.

Example command line that failed before this bug fix and now
runs:

./generalgraph --max-levels 3 --metis-agglomeration --coarse-factor 4 --num-part 30

barker29 committed:
Merge pull request #40 from LLNL/bugfix/BubbleBubbleBlock

bugfix for aggregate number in SetBubbleBubbleBlock [bugfix/BubbleBubbleBlock]

chakshinglee committed:
use subvector of global constant representation for local extension problem

barker29 commented:
It still fails:
```
{
  "finest-div-error": 0.19793223114421926,
  "finest-p-error": 0.28945323154741931,
  "finest-u-error": 0.29617999465399153,
  "operator-complexity": 1.1897781022189777
}

  (1/1) [pargraph-usegenerator] FAILED.
```
Actually all the numbers are a little different, complexity is the most different, the python expects `1.1895581044189558`.

I'm not too worried about this difference, but I would like tests to pass (just changing the `stest.py` is fine if we understand that it's a slightly different algorithm).

I would like them to be consistent, but that may not be easy. To me `1e-2` is too forgiving, I view Travis as just a rough quick test and the "real" tests run on my local machine. Ideally we would have better tests, so that tests fail only when something actually changes, and not just when some tiny numerical precision issue propagates through. On the other hand changing everything to `1e-2` seems like it will allow real bugs to escape notice.

chakshinglee commented:
Ok, this is actually not what I expected. Because on my tux box the operator complexity after the last commit is indeed `1.1895581044189558`, that means the results on our machine are different too (using the same code). I agree that we can just change the number in `stest.py`, but now I don't know which number to use as reference. And I want to know what is causing our results to be different.

For Travis, I also think `1e-2` is too loose. Does the tolerance affect the test speed that much? I thought most of the test time is spent in actual computation (coarsening, solving).

chakshinglee committed:
better names and better documentation related to 'piecewise constant' projector and ParMult between SparseMatrix and HypreParMatrix.

chakshinglee committed:
add operator complexity computation in multilevel, add a multilevel test for finite volume example

chakshinglee committed:
minor output edit

chakshinglee committed:
more documentation on 'piecewise-constant projector'

chakshinglee commented:
@barker29 I have added some documentation regarding the "piecewise-constant projection". Please let me know if they make sense (mainly in `Upscale.hpp` and `MixedMatrix.hpp`). Also, a 3-level `finitevolume` test is added. 

The OC difference in `pargraph-usegenerator` is not resolved. Since the OC results on our machines are different due to some small entries (magnitude < `1e-14`), I want to check if we have the same OC on commit 7fcdcc7. In this version the small entries are just tiny but not exactly zero in my machine, so I guess our results have a better chance to agree. My result in that version is
```
  "finest-div-error": 0.19793206480307646,
  "finest-p-error": 0.28945297306669671,
  "finest-u-error": 0.29617987947091357,
  "operator-complexity": 1.1898581014189857
``` 
If you have similar result (difference < `1e-4`), I will just use this as reference.

barker29 reviewed (approved):
I think this is ready to go now. Thanks for all the hard work!
